python/ray/util/sgd/torch/distributed_torch_runner.py

```diff
@@ -283,41 +295,55 @@ def __init__(self, *args, num_cpus=None, num_gpus=None, **kwargs):
-        self.local_device = "0"
+        self.local_cpu_device = None
+        self.local_cuda_device = "0"
         self._is_set = False
         if num_gpus:
             assert num_gpus == 1, "Does not support multi-gpu workers"
 
-        if not self.is_actor() and num_gpus > 0:
-            self._try_reserve_resources_and_set_cuda()
+        if num_cpus is None:
+            num_cpus = 0
+
+        if num_gpus is None:
+            num_gpus = 0
+
+        if not self.is_actor() and (num_cpus > 0 or num_gpus > 0):
+            self._try_reserve_and_set_resources(num_cpus, num_gpus)
 
         super(LocalDistributedRunner, self).__init__(*args, **kwargs)
 
-    def _try_reserve_resources_and_set_cuda(self, num_cpus, num_gpus):
-        visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES")
-        reserved_gpu_device = reserve_resources(num_cpus, num_gpus)
+    def _try_reserve_and_set_resources(self, num_cpus, num_gpus):
+        visible_cuda_devices = os.environ.get("CUDA_VISIBLE_DEVICES")
+        reserved_cpu_devices, reserved_cuda_device = reserve_resources(
+            num_cpus, num_gpus)
+        self._set_cpu_devices(reserved_cpu_devices)
         if num_gpus == 0:
             return
         # This needs to be set even if torch.cuda is already
         # initialized because the env var is used later when
         # starting the DDP setup.
-        os.environ["CUDA_VISIBLE_DEVICES"] = reserved_gpu_device
-        if visible_devices:
+        os.environ["CUDA_VISIBLE_DEVICES"] = reserved_cuda_device
+        if visible_cuda_devices:
             # We want to set the index on the visible devices list.
-            if reserved_gpu_device not in visible_devices:
+            if reserved_cuda_device not in visible_cuda_devices:
                 raise RuntimeError(
                     "TorchTrainer reserved a device {} that was not in the "
                     "CUDA_VISIBLE_DEVICES {}. This may be because the "
                     "Ray cluster is not set with the right env vars. "
                     "If that is not the issue, please raise a Github "
-                    "issue.".format(reserved_gpu_device, visible_devices))
-            devices = visible_devices.split(",")
-            scoped_index = devices.index(reserved_gpu_device)
+                    "issue.".format(reserved_cuda_device,
+                                    visible_cuda_devices))
+            devices = visible_cuda_devices.split(",")
+            scoped_index = devices.index(reserved_cuda_device)
             self._set_cuda_device(str(scoped_index))
         else:
             # Once cuda is initialized, torch.device ignores the os.env
             # so we have to set the right actual device.
-            self._set_cuda_device(reserved_gpu_device)
+            self._set_cuda_device(reserved_cuda_device)
+
+    def _set_cpu_devices(self, device_list):
+        # for test
+        self.local_cpu_device = device_list
```

Comment:
don't think we need to do this, see comments on the test

Reply:
removed